src/librustc/hir/check_attr.rs

```diff
@@ -53,33 +53,36 @@ impl<'a> CheckAttrVisitor<'a> {
             }
         };
         for word in words {
-            let word: &str = &word.name();
-            let message = match word {
-                "C" => {
-                    if target != Target::Struct && target != Target::Enum {
-                            "attribute should be applied to struct or enum"
-                    } else {
-                        continue
+            let name = word.name();
+            let message = match name {
+                Some(word) => match &*word {
```

Comment:
Might be able to do `match word.name().as_ref().map_or("", |s| s)` (or `map_or("", |s| &**s)`) - assuming that it has been previously validated.